### PR TITLE
feat(cli): Add cost optimization nudges for Argo CLI

### LINKED ISSUE #4548: Add cost optimization nudges on the CLI
# Summary

The nudges are currently only available on the UI (https://github.com/argoproj/argo/pull/3089). Many users may not use the UI, so add them to the CLI as well.

# Use Cases

When would you use this?

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by dinever**
Hi guys, this one closes #4548

Demo:

Screen Shot 2021-02-23 at 12 47 45 AM

This is my first contribution to Argo. Would❤️   to hear some feedback. Thanks!

Checklist:

* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

### CODE REVIEW

File: util/printer/workflow-printer.go [hunk 1]
```diff
@@ -76,16 +76,48 @@ func printTable(wfList []wfv1.Workflow, out io.Writer, opts PrintOpts) {
...
 			_, _ = fmt.Fprintf(w, "\t%d/%d/%d", pending, running, completed)
 			_, _ = fmt.Fprintf(w, "\t%s", parameterString(wf.Spec.Arguments.Parameters))
 		}
 		_, _ = fmt.Fprintf(w, "\n")
 	}
+
+	completed, incomplete := countCompletedWorkflows(wfList)
+	if completed > 100 || incomplete > 100 {
+		_, _ = fmt.Fprint(w, "\n")
+		fmt.Fprintf(w, "You have at least ")
```
Comment (simster7) [inline util/printer/workflow-printer.go:89]: Should this also disregard returned values for consistency?
```suggestion
		_, _ = fmt.Fprintf(w, "You have at least ")
```

**Review by simster7 — COMMENTED** [review]
Great! Thanks for submitting this. This looks great so far. Two things:

* Users should be able to turn off nudges. I now see that our CLI does not store any state anywhere. Not sure that we would want to add a CLI config/data store (e.g. in `~/.argo/config`) just for this, but that is a possibility. Do you have any other ideas for this?
* We have another nudge in our UI: 
https://github.com/argoproj/argo-workflows/blob/4b78a7ee4c2db9e56bf1ff0387c2de3cbe38ebf1/ui/src/app/workflows/components/workflow-details/workflow-details.tsx#L121-L123
    Could we add that as well?

**Comment by dinever** [comment]
Thanks for the review @simster7!

> Not sure that we would want to add a CLI config/data store (e.g. in ~/.argo/config) just for this, but that is a possibility.

I'm happy to implement this. However, a config data store feels like an overkill for this nudges use case unless we have other use cases that can benefit from it as well.

Another option is that we can add a `--disable-nudges` option for the CLI so that users can choose to disable it.

Let me know what you think. Thanks!

**Comment by dinever** [comment]
Hi @simster7,

Thanks for the review! I made some changes based on your feedback.

1. af3f1f656cc47779c6e2d4dcae5e21672ee0a2b3: Added a `--no-nudges` option for the CLI impacted. User can choose to not print the nudges
2. ffe487f2fb0b5829345bfa870c02a59d126adfdc: Added a security nudge for `argo get`
Screen Shot 2021-02-24 at 1 14 19 AM
3. c610490753c22b65285ab00b9b282da83f358700: Discard return values for `fmt.Print` and `fmt.Printf` for consistency

Let me know your thoughts. Thanks!

**Comment by alexec** [comment]
My thoughts on this:

1. Great stuff!
2. I'm not sure it is worth it to make it possible to disable nudges in the CLI. Instead, let's see if people ask for that feature.
3. Should we do security nudges too?
4. Nudges need to use "exec spec". I think there is `Workflow.ExecSpec()` for that.

**Comment by simster7** [comment]
> Should we do security nudges too?

I think security nudges are already implemented.

> Nudges need to use "exec spec". I think there is Workflow.ExecSpec() for that.

Yes, good point.

File: util/printer/workflow-printer.go [hunk 2]
```diff
@@ -76,16 +77,48 @@ func printTable(wfList []wfv1.Workflow, out io.Writer, opts PrintOpts) {
...
+			_, _ = fmt.Fprintf(w, "%d incomplete ", incomplete)
+		}
+		if incomplete > 100 && completed > 100 {
+			_, _ = fmt.Fprint(w, "and ")
+		}
+		if completed > 100 {
+			_, _ = fmt.Fprintf(w, "%d completed ", completed)
+		}
+		_, _ = fmt.Fprint(w, "workflows. Reducing the total number of workflows will reduce your costs.\n")
+		_, _ = fmt.Fprint(w, "Learn more at https://argoproj.github.io/argo-workflows/cost-optimisation/\n")
```
Comment (alexec) [inline util/printer/workflow-printer.go:101]: can we move the security nudge code here please?

File: util/printer/workflow-printer.go [hunk 3]
```diff
@@ -76,16 +77,48 @@ func printTable(wfList []wfv1.Workflow, out io.Writer, opts PrintOpts) {
...
+		if incomplete > 100 {
+			_, _ = fmt.Fprintf(w, "%d incomplete ", incomplete)
+		}
+		if incomplete > 100 && completed > 100 {
+			_, _ = fmt.Fprint(w, "and ")
+		}
+		if completed > 100 {
+			_, _ = fmt.Fprintf(w, "%d completed ", completed)
+		}
+		_, _ = fmt.Fprint(w, "workflows. Reducing the total number of workflows will reduce your costs.\n")
```
Comment (alexec) [inline util/printer/workflow-printer.go:100]: minor - could you use `Fprintln`?

**Review by alexec — CHANGES_REQUESTED** [review]

**Review by alexec — APPROVED** [review]

**Pull request merged**